tailfs/shared/readonlydir.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// package shared contains components shared by different tailfs packages
+package shared
+
+import (
+	"errors"
+	"io"
+	"io/fs"
+	"sync"
+)
+
+// DirFile implements webdav.File for a virtual directory.
+// It mimics the behavior of an os.File that is pointing at a real directory.
+type DirFile struct {
+	// Info provides the fs.FileInfo for this directory
+	Info fs.FileInfo
```

Comment:
Because the DirFile is constructed with an Info, like so

```
&shared.DirFile{
		Info: fi,
		LoadChildren: func() ([]fs.FileInfo, error) {
			ctxWithTimeout, cancel := context.WithTimeout(context.Background(), opTimeout)
			defer cancel()

			dirInfos, err := wfs.Client.ReadDir(ctxWithTimeout, name)
			if err != nil {
				wfs.logf("encountered error reading children of '%v', returning empty list: %v", name, err)
				// We do not return the actual error here because some WebDAV clients
				// will take that as an invitation to retry, hanging in the process.
				return dirInfos, nil
			}
			if wfs.statCache != nil {
				wfs.statCache.set(name, dirInfos)
			}
			return dirInfos, nil
		},
	}
```